tools/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/Transaction.java

```diff
@@ -472,28 +473,40 @@ public WildcardRDFRemover(RepositoryConnection conn) {
 
 		@Override
 		public void handleStatement(Statement st) throws RDFHandlerException {
+			System.out.println(st);
+			Thread.dumpStack();
 			Resource subject = SESAME.WILDCARD.equals(st.getSubject()) ? null : st.getSubject();
 			IRI predicate = SESAME.WILDCARD.equals(st.getPredicate()) ? null : st.getPredicate();
 			Value object = SESAME.WILDCARD.equals(st.getObject()) ? null : st.getObject();
 
-			// use the RepositoryConnection.clear operation if we're removing
-			// all statements
-			final boolean clearAllTriples = subject == null && predicate == null && object == null;
+			// use the RepositoryConnection.clear operation if we're removing all statements
+			boolean clearAllTriples = subject == null && predicate == null && object == null;
+
+			boolean allContexts = st.getContext() == null;
+			boolean defaultContext = RDF4J.NIL.equals(st.getContext());
 
 			try {
-				Resource context = st.getContext();
-				if (context != null) {
+				if (allContexts) {
 					if (clearAllTriples) {
-						conn.clear(context);
+						System.out.println("conn.clear();");
+						conn.clear();
 					} else {
-						conn.remove(subject, predicate, object, context);
+						System.out.println("conn.remove(subject, predicate, object);");
+						conn.remove(subject, predicate, object);
 					}
-				} else {
+				} else if (defaultContext) {
 					if (clearAllTriples) {
-						conn.clear();
+						System.out.println("conn.clear(((Resource) null));");
+						conn.clear(((Resource) null));
 					} else {
-						conn.remove(subject, predicate, object);
+						System.out.println("conn.remove(subject, predicate, object, ((Resource) null));");
+						conn.remove(subject, predicate, object, ((Resource) null));
 					}
+				} else {
+					assert st.getContext() != null;
+					System.out.println("conn.remove(" + subject + ", " + predicate + ", " + object + ", "
+							+ st.getContext() + ");");
+					conn.remove(subject, predicate, object, st.getContext());
 				}
```

Comment:
I don't really like the complexity of the branching logic here - so many nested if-elses and boolean flags. I think if we separate out the logic that determines what the value of the context param should be into a separate method, we can simplify this a bit.

Reply:
Yeah. This is very much work-in-progress. Since I can't debug the code directly I've made it particularly explicit and verbose because I don't trust myself not to make stupid mistakes :P 

I'll see how I can simplify it once I've got all the bugs ironed out.